### PR TITLE
PUB-737 Build Subscriptions Pending Submission Screen(s)

### PR DESCRIPTION
**Pull request by marcogagliardi**
### JIRA link ###
https://tools.hmcts.net/jira/browse/PUB-737

### Change description ###
Implement pending submissions screens which utilises `Redis` cache in order to add multiple subscriptions.
Subscriptions are divided into court or case subscriptions.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```
nopending
pendingsubscriptions
image

### CODE REVIEW

File: src/main/views/subscription-confirmed.njk [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{% extends "template.njk" %}
```
Comment (ChrisS1512) [inline src/main/views/subscription-confirmed.njk:1]: This page looks a bit different from the one in the JIRA ticket. Just want to double check the below is correct (titles, bullet points)?

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/controllers/common/utils.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import {cloneDeep} from 'lodash';
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply src/main/controllers/common/utils.ts]: done

File: src/main/views/subscription-confirmed.njk (hunk 1 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmed.njk:1]: I have taken the screen from the wireframe on herokuapp

**Comment by commitBlob** [comment]
You should merge to pre-release instead of master

**Comment by marcogagliardi** [comment]
> You should merge to pre-release instead of master
this is a branch of pre-release

File: src/main/controllers/SubscriptionConfirmedController.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { Response } from 'express';
+import {PipRequest} from '../models/request/PipRequest';
+import {cloneDeep} from 'lodash';
+
+export default class SubscriptionConfirmedController {
+
+  public async get(req: PipRequest, res: Response): Promise<void> {
+
+    res.render('subscription-confirmed', {
+      ...cloneDeep(req.i18n.getDataByLanguage(req.lng)['subscription-confirmed']),
```
Comment (kurt-uk) [inline src/main/controllers/SubscriptionConfirmedController.ts:10]: with no other options provided you can just call this inline, theres no need to deepen the params
`res.render('subscription-confirmed', req.i18n.getDataByLanguage(req.lng)['subscription-confirmed']);`

File: src/main/controllers/SubscriptionUrnSearchResultController.ts [hunk 4]
```diff
@@ -1,32 +1,28 @@
...
+      validateRendering(searchResults,'subscription-urn-search-results',req, res, searchInput);
 
     }
     else {
       res.render('error', req.i18n.getDataByLanguage(req.lng).error);
     }
   }
+
+  public async post(req: PipRequest, res: Response): Promise<void> {
+    res.redirect(`subscription-confirmation?search-input=${searchInput}&stype=urn`);
```
Comment (kurt-uk) [inline src/main/controllers/SubscriptionUrnSearchResultController.ts:25]: can this live in the forms post if this is all thats needed?

File: src/main/resources/locales/en/subscription-urn-search.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "title": "Subscription URN search",
-  "header": "Enter a unique reference number",
+  "header": "What is the unique reference number (URN)?",
```
Comment (kurt-uk) [inline src/main/resources/locales/en/subscription-urn-search.json:3]: It looks like it was right the first time based on the screenshots from Nige

File: src/main/controllers/SubscriptionUrnSearchResultController.ts [hunk 6]
```diff
@@ -1,32 +1,28 @@
...
-      if (searchResults) {
-        res.render('subscription-urn-search-results', {
-          ...cloneDeep(req.i18n.getDataByLanguage(req.lng)['subscription-urn-search-results']),
-          searchInput : searchInput,
-          searchResults: searchResults,
-        });
-      } else {
-        res.render('error', req.i18n.getDataByLanguage(req.lng).error);
-      }
+      validateRendering(searchResults,'subscription-urn-search-results',req, res, searchInput);
```
Comment (kurt-uk) [inline src/main/controllers/SubscriptionUrnSearchResultController.ts:16]: where is the error state screens? (not the error page)

File: src/main/views/subscription-confirmation.njk [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      <tr class="govuk-table__row">
+        <td scope="row" class="govuk-table__cell--no-line">{{ searchResults.urn }}</td>
+        <td scope="row" class="govuk-table__cell--no-line">{{ searchResults.caseNumber }}</td>
+        <td class="govuk-table__cell--no-line govuk-!-text-align-right">{{ searchResults.caseName }}</td>
+        <td class="govuk-table__cell--no-line govuk-!-text-align-right"><a href="#">{{ removeText }}</a></td>
+      </tr>
+
+    </tbody>
+  </table>
+  {{ submitButton(buttonText) }}
```
Comment (kurt-uk) [inline src/main/views/subscription-confirmation.njk:45]: youre missing the functionality to add more subscriptions as per ticket https://tools.hmcts.net/jira/browse/PUB-737

File: src/main/views/subscription-confirmed.njk (hunk 1 above)
Comment (kurt-uk) [reply src/main/views/subscription-confirmed.njk:1]: heroku should be used as a lose guide, the standard we should be using is the AC on the ticket and the screenshots provided in the ticket, Heroku is used to prototype different designs, the approved and needing developing designs are the ones provided on the ticket, 9 times out of 10 its the ticket we should be using unless stated otherwise

**Review by kurt-uk — CHANGES_REQUESTED** [review]
i cant see on any of your screenshots the ability to continue adding, this will need to be added too

File: src/main/resources/locales/en/subscription-urn-search.json (hunk 5 above)
Comment (marcogagliardi) [reply src/main/resources/locales/en/subscription-urn-search.json:3]: I have taken the text from the herokuapp portal

File: src/main/controllers/SubscriptionUrnSearchResultController.ts (hunk 6 above)
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionUrnSearchResultController.ts:16]: not sure what you mean

File: src/main/views/subscription-confirmation.njk (hunk 7 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmation.njk:45]: Nigel said we didn't have it yet

File: src/main/controllers/SubscriptionConfirmationController.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import {SubscriptionService} from '../service/subscriptionService';
+import validateRendering from '../common/utils';
+
+const subscriptionService = new SubscriptionService();
+let searchInput;
+export default class SubscriptionConfirmationController {
+
+  public async get(req: PipRequest, res: Response): Promise<void> {
+    searchInput = req.query['search-input'];
+    const type = req.query['stype'];
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionConfirmationController.ts:12]: it is not a typo. stype => search type

File: src/main/controllers/SubscriptionConfirmedController.ts (hunk 3 above)
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionConfirmedController.ts:10]: this has been refactored

File: src/main/controllers/SubscriptionUrnSearchResultController.ts (hunk 4 above)
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionUrnSearchResultController.ts:31]: this has been refactored

File: src/main/views/subscription-confirmation.njk (hunk 7 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmation.njk:45]: the add button has been added

File: src/main/common/utils.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import {cloneDeep} from 'lodash';
+
+export default function validateRendering(searchResults,path, req, res, searchInput): void {
+  if (searchResults) {
+    res.render(path, {
+      ...cloneDeep(req.i18n.getDataByLanguage(req.lng)[path]),
+      searchInput : searchInput,
+      searchResults: searchResults,
```
Comment (junaidiqbalmoj) [inline src/main/common/utils.ts:8]: you do not need to assign here, you can do like this in typescript:

searchInput,
searchResults,

File: src/main/views/subscription-confirmation.njk (hunk 7 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmation.njk:45]: add link added

File: src/main/controllers/SubscriptionUrnSearchResultController.ts (hunk 4 above)
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionUrnSearchResultController.ts:31]: refactored

File: src/main/views/subscription-confirmed.njk (hunk 1 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmed.njk:1]: spoke to Nigel. He will add comments in the ticket

**Comment by marcogagliardi** [comment]
> i cant see on any of your screenshots the ability to continue adding, this will need to be added too

link to add added

File: src/main/common/utils.ts (hunk 9 above)
Comment (marcogagliardi) [reply src/main/common/utils.ts]: done

**Comment by alberto-hortiguela** [comment]
Hi Marco, please fix these sections below.
image
image

File: src/main/service/subscriptionService.ts [hunk 10]
```diff
@@ -66,4 +68,23 @@ export class SubscriptionService {
     }
   }
 
+  public setPendingSubscriptions(searchResult, user): boolean {
+    pendingSubscriptionsFromCache.setPendingSubscriptions(searchResult, user);
+    return true;
```
Comment (junaidiqbalmoj) [inline src/main/service/subscriptionService.ts:73]: This function will always return "true". Is it correct?

File: src/main/views/subscription-confirmation.njk [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+{% block beforeContent %}
+  {{ super() }}
+  {{ govukBackLink({
+    text: text,
+    href: '#',
+    attributes: {
+      onClick: "history.back();event.preventDefault();"
+    }
+  }) }}
```
Comment (junaidiqbalmoj) [inline src/main/views/subscription-confirmation.njk:18]: can you use {{ goBack() }} here?

File: src/main/resources/requests/subscriptionRequests.ts [hunk 12]
```diff
@@ -3,6 +3,7 @@ import path from 'path';
 import {dataManagementApi} from './utils/axiosConfig';
 import {CaseSubscription} from '../../models/caseSubscription';
 
+
```
Comment (commitBlob) [inline src/main/resources/requests/subscriptionRequests.ts:6]: how much value this extra line adds?

File: src/main/resources/requests/subscriptionRequests.ts [hunk 13]
```diff
@@ -33,4 +34,15 @@ export class SubscriptionRequests {
     }
     return null;
   }
+
+  public async subscribe(searchResult: Array<CaseSubscription>, user): Promise<boolean> {
```
Comment (commitBlob) [inline src/main/resources/requests/subscriptionRequests.ts:38]: Since this function doesn't do anything at the moment use this shorthand.
return !!(user); // this returns boolean

File: src/main/resources/locales/en/subscription-urn-search.json (hunk 5 above)
Comment (kurt-uk) [reply src/main/resources/locales/en/subscription-urn-search.json:3]: we should be using the screenshots on the ticket unless stated otherwise to take from heroku, heroku is used for prototyping and therefore will be different in some content as different things will be being trialled, so as default we should be taking the approved screenshots from the ticket

File: src/main/resources/requests/hearingRequests.ts [hunk 14]
```diff
@@ -18,5 +24,30 @@ export class HearingRequests {
     }
     return [];
   }
+
+  public async getHearingsById(id: number): Promise<Hearing> {
```
Comment (commitBlob) [inline src/main/resources/requests/hearingRequests.ts:28]: why do we need this?

File: src/main/controllers/CaseNameSearchResultsController.ts [hunk 15]
```diff
@@ -18,4 +20,29 @@ export default class CaseNameSearchResultsController {
       res.render('error', req.i18n.getDataByLanguage(req.lng).error);
     }
   }
+
+  public async post(req: PipRequest, res: Response): Promise<void> {
```
Comment (commitBlob) [inline src/main/controllers/CaseNameSearchResultsController.ts:24]: why has this been added? caseNameSearchResultsController used to post data to pending-subscriptions page

File: src/main/views/case-name-search-results.njk [hunk 16]
```diff
@@ -1,55 +1,55 @@
...
+    {{ title }}
+{% endblock %}
+{% block beforeContent %}
+    {{ super() }}
+    {{ goBack() }}
+{% endblock %}
+{% block content %}
+    <h1 class="govuk-heading-l">{{ heading }}</h1>
+    <p class="govuk-body govuk-!-font-weight-bold">{{ searchResults | length }} {{ resultsCount }}</p>
+    <form method="post">
```
Comment (commitBlob) [inline src/main/views/case-name-search-results.njk:16]: why did you remove pending-subscriptions post?

File: src/main/views/court-name-search.njk [hunk 17]
```diff
@@ -1,167 +1,167 @@
...
+                        },
+                        categories: checkedFilters
+                    },
+                    optionsHtml: filterOptionsHtml
+                }) }}
+            </form>
+        </div>
+
+        <div class="layout-width-three-fifths">
+            <form class="moj-action-bar" action="subscription-confirmation" method="post">
```
Comment (commitBlob) [inline src/main/views/court-name-search.njk:84]: why did you remove pending-subscriptions from here?

File: src/main/views/subscription-confirmation.njk (hunk 11 above)
Comment (marcogagliardi) [reply src/main/views/subscription-confirmation.njk:18]: back button has been removed. Please see comment from Nigel in the ticket

File: src/main/views/subscription-urn-search-results.njk [hunk 18]
```diff
@@ -1,13 +1,14 @@
...
 {% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "./macros/common-components.njk" import submitButton %}
 
 {% block pageTitle %}
   {{ title }}
 {% endblock %}
 
 {% block beforeContent %}
+  {{ super() }}
   {{ govukBackLink({
```
Comment (commitBlob) [inline src/main/views/subscription-urn-search-results.njk:12]: can we use back from the common components here?

File: src/main/views/subscription-urn-search.njk [hunk 19]
```diff
@@ -9,13 +9,15 @@
...
 {% block beforeContent %}
-{{ govukBackLink({
-  text: text,
-  href: '#',
-  attributes: {
-    onClick: "history.back();event.preventDefault();"
-  }
-}) }}
+  {{ super() }}
+  {{ govukBackLink({
```
Comment (commitBlob) [inline src/main/views/subscription-urn-search.njk:13]: common components back please

File: src/main/service/subscriptionService.ts (hunk 10 above)
Comment (marcogagliardi) [reply src/main/service/subscriptionService.ts:73]: refactored

File: src/test/unit/controllers/SubscriptionConfirmationController.test.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+      'subscription-confirmation': {},
+    };
+
+    const response = {
+      render: function () {
+        return '';
+      },
+    } as unknown as Response;
+    const request = mockRequest(i18n);
+    //request.query = { 'search-input': '123456789', 'stype': 'urn'};
```
Comment (commitBlob) [inline src/test/unit/controllers/SubscriptionConfirmationController.test.ts:28]: re move commented out lines please

File: src/test/unit/mocks/caseHearings.json [hunk 21]
```diff
@@ -0,0 +1,25 @@
+ [
```
Comment (commitBlob) [inline src/test/unit/mocks/caseHearings.json:1]: you created this already

File: src/test/unit/views/subscription-confirmation.test.ts [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+
+const rawData = fs.readFileSync(path.resolve(__dirname, '../mocks/subscriptionListResult.json'), 'utf-8');
+const subscriptionsData = JSON.parse(rawData);
+const rawData2 = fs.readFileSync(path.resolve(__dirname, '../mocks/caseHearings.json'), 'utf-8');
+const subscriptionsData2 = JSON.parse(rawData2);
+sinon.stub(SubscriptionRequests.prototype, 'getSubscriptionByUrn').returns(subscriptionsData);
+sinon.stub(PendingSubscriptionsFromCache.prototype, 'getPendingSubscriptions').returns(subscriptionsData2);
+
+
+jest.mock('axios', () => {
```
Comment (commitBlob) [inline src/test/unit/views/subscription-confirmation.test.ts:27]: use sinon to mock this

**Review by commitBlob — CHANGES_REQUESTED** [review]
Please stick to guidelines we agreed on the Github discussion
https://cjscommonplatform.sharepoint.com/:w:/r/sites/PublishingInformationHub/_layouts/15/Doc.aspx?sourcedoc=%7BDE79524F-BE38-4ECC-A17F-AEC6FEAF60A0%7D&file=GitHub%20repo%20%26%20PR%20improvements.docx&action=default

File: src/main/controllers/SubscriptionConfirmationController.ts [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+    validateRendering(searchResults,'subscription-confirmation',req, res, null);
+
+  }
+
+  public async post(req: PipRequest, res: Response): Promise<void> {
+    const searchResults = await subscriptionService.getPendingSubscriptions(req.user);
+    await subscriptionService.subscribe(searchResults, req.user);
+    res.render('subscription-confirmed', {
+      ...cloneDeep(req.i18n.getDataByLanguage(req.lng)['subscription-confirmed']),
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionConfirmationController.ts:21]: ok removed

File: src/main/resources/locales/en/subscription-confirmed.json [hunk 24]
```diff
@@ -0,0 +1,13 @@
+{
+  "title": "Courts and tribunals listing - GOV.UK",
+  "header": "Subscription confirmed",
+  "subheader": "Your subscription has been successful.",
+  "p1": "To continue, you can go to ",
+  "link_text": "your account",
+  "link": "#",
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply src/main/resources/locales/en/subscription-confirmed.json:7]: changed to /my-account page

File: src/main/resources/locales/en/subscription-confirmed.json [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+  "p2": " in order to:",
+  "list": ["add a new subscription","manage your current subscriptions","find a court or tribunal"],
+  "p3_link": "#",
+  "p3_link_text": "What did you think about this service?",
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply src/main/resources/locales/en/subscription-confirmed.json:11]: yes. I will remove from the locales

File: src/main/views/subscription-urn-search-results.njk (hunk 18 above)
Comment (marcogagliardi) [reply src/main/views/subscription-urn-search-results.njk:12]: refactored

File: src/test/unit/controllers/SubscriptionConfirmationController.test.ts (hunk 20 above)
Comment (marcogagliardi) [reply src/test/unit/controllers/SubscriptionConfirmationController.test.ts:28]: comment removed

File: src/test/unit/views/subscription-confirmation.test.ts (hunk 22 above)
Comment (marcogagliardi) [reply src/test/unit/views/subscription-confirmation.test.ts:27]: sorry I can't understand

File: src/test/unit/mocks/caseHearings.json (hunk 21 above)
Comment (marcogagliardi) [reply src/test/unit/mocks/caseHearings.json:1]: removed

**Review by commitBlob — CHANGES_REQUESTED** [review]
I also believe that subscription-confirmation and subscription-confirmed pages and controllers need refactoring. More appropriate name for the first one would be pending-subscriptions as that page is supposed to contain all subscriptions user selected but has not subscribed to yet.
Success page should be subscription-confirmation

File: src/main/controllers/SubscriptionConfirmedController.ts (hunk 3 above)
Comment (marcogagliardi) [reply src/main/controllers/SubscriptionConfirmedController.ts:10]: this has been refactored

Comment (marcogagliardi) [reply src/main/controllers/SubscriptionConfirmedController.ts:10]: this has been refactored

File: src/main/views/subscription-urn-search.njk (hunk 19 above)
Comment (marcogagliardi) [reply src/main/views/subscription-urn-search.njk:13]: this has been refactored

File: package.json [hunk 26]
```diff
@@ -62,6 +62,7 @@
     "i18next-http-middleware": "^3.1.1",
     "ioredis": "^4.28.0",
     "jquery": "^3.5.1",
+    "json-schema": "^0.4.0",
```
[the remark this answers is not in the document]
Comment (marcogagliardi) [reply package.json]: removed

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]
I'm getting a couple of issues when trying this out.

1) If I try and add a subscription by URN, clicking Continue on the on the Subscription URN search results screen gives me an Something went wrong error

2) If I go down the Subscription Add via Court Name search route, I get a Subscription Confirmed screen instantly when selecting a court and clicking continue.

Are you getting the same issues as above? I'm wondering if it's my redis setup

File: src/main/resources/locales/en/subscription-urn-search.json (hunk 5 above)
Comment (kurt-uk) [reply src/main/resources/locales/en/subscription-urn-search.json:3]: unresolving as i think you must have accidentally resolved

File: src/main/resources/requests/hearingRequests.ts (hunk 14 above)
Comment (commitBlob) [reply src/main/resources/requests/hearingRequests.ts:28]: resolving as it seems like we need to get hearingsby id

File: src/main/views/pending-subscriptions.njk [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+        <td scope="row" class="govuk-table__cell--no-line">{{ hearing.urn }}</td>
+        <td scope="row" class="govuk-table__cell--no-line">{{ hearing.caseNumber }}</td>
+        <td class="govuk-table__cell--no-line govuk-!-text-align-right">{{ hearing.caseName }}</td>
+        <td class="govuk-table__cell--no-line govuk-!-text-align-right"><a href="/remove-case?id={{ hearing.hearingId }}">{{ removeText }}</a></td>
+      </tr>
+    {% endfor %}
+    </tbody>
+  </table>
+  {{ submitButton(buttonText) }}
+  <a class="govuk-link" href="{{ addOther_link }}">{{ addOther_text }}</a>
```
Comment (ChrisS1512) [inline src/main/views/pending-subscriptions.njk:46]: The anchor tag needs updating to be centred to the right of the button (See pending-subscriptions.png screenshot in ticket). Currently it is displayed aligned to the top.

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

**Comment by marcogagliardi** [comment]
As for Nigel comments on the ticket the another link will be dealt in different ticket

**Comment by furnivall** [comment]
> 

test

File: src/main/views/pending-subscriptions.njk (hunk 27 above)
Comment (furnivall) [reply src/main/views/pending-subscriptions.njk]: test

File: src/main/models/caseSubscription.ts [hunk 28]
```diff
@@ -1,11 +1,19 @@
 export interface CaseSubscription {
```
Comment (ChrisS1512) [inline src/main/models/caseSubscription.ts:1]: My understanding is subscriptions aren't specific to a hearing, and are just general search criteria based on a caseName, caseNumber and urn that is then stored as a subscription, therefore I'm not sure we need hearing ID, date / judge / platform / court ID. Would be good to discuss, as I could be wrong on this.

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/models/caseSubscription.ts (hunk 28 above)
Comment (ChrisS1512) [reply src/main/models/caseSubscription.ts:1]: Following a discussion, I have raised the investigation into the granularity of subscriptions into a separate ticket: https://tools.hmcts.net/jira/browse/PUB-954.

**Review by ChrisS1512 — APPROVED** [review]

File: src/main/models/caseSubscription.ts (hunk 28 above)
Comment (kurt-uk) [reply src/main/models/caseSubscription.ts:1]: this model is a hearing however, just duplicated for some reason with another name

File: src/main/resources/requests/hearingRequests.ts [hunk 29]
```diff
@@ -19,6 +19,22 @@ export class HearingRequests {
...
+        console.log(error.response.data);
+      } else if (error.request) {
+        console.log(`Request failed. ${error.request}`);
+      } else {
+        console.log(`ERROR: ${error.message}`);
+      }
+    }
+    return null;
+  }
+
```
Comment (kurt-uk) [inline src/main/resources/requests/hearingRequests.ts:37]: how is this method different to the one below?

File: src/main/service/hearingService.ts [hunk 30]
```diff
@@ -8,6 +8,10 @@ export class HearingService {
     return await hearingRequests.getHearingsByCaseName(searchQuery);
   }
 
+  public async getCaseByNumber(caseNumber): Promise<Hearing> {
+    return await hearingRequests.getCaseByCaseNumber(caseNumber);
+  }
+
```
Comment (kurt-uk) [inline src/main/service/hearingService.ts:14]: again is this needed?

File: src/main/service/subscriptionService.ts [hunk 31]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
+    const selectionList = Object.keys(pendingSubscription);
+    for (const selectionName of selectionList) {
+      let hearingIdsList = [];
+      let courtIdsList = [];
+      let caseDetailsList: Hearing[] = [];
+      let courtDetailsList: Court[] = [];
+      switch (selectionName) {
+        case 'case-number':
+        case 'hearing-selections[]':
+          // pendingSubscription.selectionName gives undefined
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:86]: can yeet this comment out

File: src/main/service/subscriptionService.ts [hunk 32]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
 
+  public async handleNewSubscription(pendingSubscription, user): Promise<void> {
+    const selectionList = Object.keys(pendingSubscription);
+    for (const selectionName of selectionList) {
+      let hearingIdsList = [];
+      let courtIdsList = [];
+      let caseDetailsList: Hearing[] = [];
+      let courtDetailsList: Court[] = [];
+      switch (selectionName) {
+        case 'case-number':
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:84]: what about urn case?

File: src/main/service/subscriptionService.ts [hunk 33]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
+    }
+    return subscribed;
+  }
+
+  createSubscriptionPayload(pendingSubscription, subscriptionType, userId): object {
+    let payload;
+    switch (subscriptionType) {
+      case 'courts':
+        payload = {
+          channel: 'API',
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:162]: How come this is API? should it not be Email?

File: src/main/service/subscriptionService.ts [hunk 34]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
+          searchType: 'COURT_ID',
+          searchValue: pendingSubscription.courtId,
+          userId,
+        };
+        break;
+      case 'cases':
+        payload = {
+          channel: 'API',
+          searchType: 'CASE_ID',
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:171]: youve not allowed the URN search type

File: src/main/service/subscriptionService.ts [hunk 35]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
+  }
+
+  public async getPendingSubscriptions(userId, subscriptionType): Promise<any[]> {
+    return await pendingSubscriptionsFromCache.getPendingSubscriptions(userId, subscriptionType);
+  }
+
+  public async subscribe(userId): Promise<boolean> {
+    let subscribed = true;
+    const casesType = 'cases';
+    const courtsType = 'courts';
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:139]: these can live at class level?

File: src/main/service/subscriptionService.ts [hunk 36]
```diff
@@ -66,4 +73,119 @@ export class SubscriptionService {
...
+          searchType: 'COURT_ID',
+          searchValue: pendingSubscription.courtId,
+          userId,
+        };
+        break;
+      case 'cases':
+        payload = {
+          channel: 'API',
+          searchType: 'CASE_ID',
+          searchValue: pendingSubscription.caseNumber,
```
Comment (kurt-uk) [inline src/main/service/subscriptionService.ts:172]: same witht he search value, need to accommodate URN number

File: src/test/unit/service/subscriptionService.test.ts [hunk 37]
```diff
@@ -39,5 +108,144 @@ describe('generate rows functions with subscriptions', () => {
...
+});
+
+describe('handleNewSubscription function', () => {
+  it('should add new case subscription', async () => {
+    const pendingSubscription = {'hearing-selections[]': 'T485914'};
+    await subscriptionService.handleNewSubscription(pendingSubscription, '1');
+  });
+
+  it('should add new case subscriptions', async () => {
+    const pendingSubscription = {'hearing-selections[]': ['T485914', 'T485912']};
```
Comment (kurt-uk) [inline src/test/unit/service/subscriptionService.test.ts:120]: need some tests that have `case-number` and `hearing-selections[]`

File: src/main/resources/requests/hearingRequests.ts (hunk 29 above)
Comment (commitBlob) [reply src/main/resources/requests/hearingRequests.ts:37]: duplicate code, refactored

File: src/main/service/hearingService.ts (hunk 30 above)
Comment (commitBlob) [reply src/main/service/hearingService.ts:14]: removed as a part of above refactor

File: src/main/service/subscriptionService.ts (hunk 31 above)
Comment (commitBlob) [reply src/main/service/subscriptionService.ts:86]: might be worth keeping as it is a reference to approach taken on the line bellow

File: src/main/service/subscriptionService.ts (hunk 35 above)
Comment (commitBlob) [reply src/main/service/subscriptionService.ts:139]: It is only used within that function so I thought it might be better to keep it scoped?

File: src/main/service/subscriptionService.ts (hunk 33 above)
Comment (commitBlob) [reply src/main/service/subscriptionService.ts]: Was my misunderstanding, refactored to EMAIL

File: .eslintrc.js [hunk 38]
```diff
@@ -13,6 +13,7 @@ module.exports = {
         "quotes": ["error", "single"],
         "comma-dangle": ["error", "always-multiline"],
         "semi": ["error", "always"],
+        "no-multiple-empty-lines": ["error", {"max": 1}]
```
Comment (kurt-uk) [inline .eslintrc.js:16]: image

File: src/main/service/subscriptionService.ts (hunk 32 above)
Comment (commitBlob) [reply src/main/service/subscriptionService.ts:84]: added urn case

File: src/test/unit/service/subscriptionService.test.ts (hunk 37 above)
Comment (commitBlob) [reply src/test/unit/service/subscriptionService.test.ts:130]: there are no cases where body data will be both, it will be only one of these 3:
- urn
- hearing-selections[]
- court-selections[]

**Review by kurt-uk — DISMISSED** [review]

File: src/main/models/caseSubscription.ts (hunk 28 above)
Comment (commitBlob) [reply src/main/models/caseSubscription.ts]: Resolved as this has been removed and hearing model is used instead

**Review by commitBlob — DISMISSED** [review]

**Review by kurt-uk — APPROVED** [review]

**Review by kurt-uk — APPROVED** [review]

**Review by kurt-uk — APPROVED** [review]

**Review by commitBlob — APPROVED** [review]
